### PR TITLE
Added --name argument 

### PR DESCRIPTION
**Pull request by acharles7**
## Summary
Refer #119
* Added `--name` argument to fixit's add_new_rule CLI
`python -m fixit.cli.add_new_rule --path fixit/rules/my_rule.py --name abcd`

`--file` = `fixit/rules/my_rule.py`
`--name` = `abcd`

The file will be created at `fixit/rules/my_rule.py` with class name  `abcd` as `AbcdRule` 

```python
class AbcdRule(CstLintRule):
    """
    docstring or new_rule description
    """

    MESSAGE = "Enter rule description message"

    VALID = [Valid("'example'")]

    INVALID = [Invalid("'example'")]

```
* Added `snake_to_camelcase` function to `fixit/cli/utils.py`
* Added documentation to `docs/source/build_a_lint_rule.ipynb`

### CODE REVIEW

**Comment by acharles7** [comment]
Please have a look at the docs builds error. 
We have to avoid using `In[2]` to write in `examples.py`

**Comment by acharles7** [comment]
@jimmylai Can you please look into it?

File: fixit/cli/add_new_rule.py [hunk 1]
```diff
@@ -64,10 +65,19 @@ def is_path_exists(path: str) -> Path:
         return filepath
 
 
-def create_rule_file(file_path: Path) -> None:
+def is_valid_name(name: str) -> str:
+    """Check for valid rule name """
+    if name.endswith(("Rule", "Rules", "rule", "rules")):
```
Comment (isidentical) [inline fixit/cli/add_new_rule.py:70]:
```suggestion
    if name.casefold().endswith(("rule", "rules")):
```

File: fixit/cli/add_new_rule.py [hunk 2]
```diff
@@ -64,10 +65,19 @@ def is_path_exists(path: str) -> Path:
...
+        raise NameError("Please enter rule name without the suffix, `rule` or `Rule`")
```
Comment (isidentical) [inline fixit/cli/add_new_rule.py:71]: It would be more consistent if you use single quotes instead of backticks (just like KeyError etc.)

Comment (isidentical) [reply fixit/cli/add_new_rule.py:71]: Also, NameError doesn't suit well, maybe use ValueError?

File: fixit/cli/add_new_rule.py [hunk 3]
```diff
@@ -64,10 +65,19 @@ def is_path_exists(path: str) -> Path:
...
+    return snake_to_camelcase(name) if name else ""
```
Comment (isidentical) [inline fixit/cli/add_new_rule.py:72]: You can omit `if name else ""` since `snake_to_camelcase` with empty string would result with an empty string.

**Review by isidentical — COMMENTED** [review]
Just suggestion, but how would it look using the path (like `chained_isinstance_check.py`) when there is no name supplied (convert the path's stem into `ChainedIsinstanceCheck`)

File: fixit/cli/add_new_rule.py (hunk 1 above)
Comment (acharles7) [reply fixit/cli/add_new_rule.py]: Done

**Comment by acharles7** [comment]
> Just suggestion, but how would it look using the path (like `chained_isinstance_check.py`) when there is no name supplied (convert the path's stem into `ChainedIsinstanceCheck`)

Done. Nice idea
Thanks

File: fixit/cli/add_new_rule.py (hunk 3 above)
Comment (acharles7) [reply fixit/cli/add_new_rule.py]: done

**Comment by jimmylai** [comment]
> Please have a look at the docs builds error.
> We have to avoid using `In[2]` to write in `examples.py`

For some reason, this error just show up recently.
In[2] is a way to access the content of cell number 2 in the notebook to reuse the code example.
`tox -e docs` run successfully in my local dev. It seems specific to the CI build.
Are you able to reproduce this error locally?

File: docs/source/build_a_lint_rule.ipynb [hunk 4]
```diff
@@ -373,9 +380,9 @@
...
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.7.3"
+   "version": "3.8.2"
   }
  },
  "nbformat": 4,
  "nbformat_minor": 4
-}
+}
```
[the remark this answers is not in the document]
Comment (acharles7) [reply docs/source/build_a_lint_rule.ipynb]: done

**Comment by acharles7** [comment]
> > Please have a look at the docs builds error.
> > We have to avoid using `In[2]` to write in `examples.py`
> 
> For some reason, this error just show up recently.
> In[2] is a way to access the content of cell number 2 in the notebook to reuse the code example.
> `tox -e docs` run successfully in my local dev. It seems specific to the CI build.
> Are you able to reproduce this error locally?

I am also getting this error in local dev.

**Comment by acharles7** [comment]
We can use something like this.
```python
file_content = """\
class C:
...content of In[2]
"""
```
then write `file_content` to examples.py

File: docs/source/getting_started.ipynb [hunk 5]
```diff
@@ -64,6 +64,32 @@
...
+    "        filtered_char = []\n",
+    "        for char in self.attr:\n",
+    "            if char is not \"a\":\n",
+    "                filtered_char.append(char)\n",
+    "\n",
+    "        index_to_char = dict([(idx, char) for idx, char in enumerate(filtered_char)])\n",
+    "        return index_to_char\n",
+    "\"\"\""
+   ]
+  },
```
Comment (jimmylai) [inline docs/source/getting_started.ipynb:92]: Copy the example code is not ideal. It'll make the future updated less easy.

I've find a workaround in https://github.com/Instagram/Fixit/pull/133

Comment (acharles7) [reply docs/source/getting_started.ipynb:92]: I'll revert it. Thanks

**Review by jimmylai — APPROVED** [review]
Other than comments, LGTM.
Thanks for contributing!

File: fixit/cli/add_new_rule.py [hunk 6]
```diff
@@ -83,12 +94,23 @@ def main() -> None:
...
+        default=Path("fixit/rules/new.py"),
+        help="Path to add rule file, Default is fixit/rules/new.py",
+    )
+
+    parser.add_argument(
+        "-n",
+        "--name",
+        type=str,
+        default="",
+        help="Name of the rule, Default is New",
```
[the remark this answers is not in the document]
Comment (acharles7) [reply fixit/cli/add_new_rule.py]: done

File: docs/source/build_a_lint_rule.ipynb [hunk 7]
```diff
@@ -132,10 +132,17 @@
...
+    "    $ python -m fixit.cli.add_new_rule # Creates new.py at fixit/rules/new.py\n",
+    "    $ python -m fixit.cli.add_new_rule --path fixit/rules/my_rule.py --name rule_name # Creates my_rule.py at path specified \n",
     "\n",
-    "This will generate a rule file used to create and add new rule to fixit module."
+    "This will generate a rule file used to create and add new rule to fixit module. \n",
+    "\n",
+    "The `fixit.cli.add_new_rule` contains two argument, ``-path`` and ``--name``\n",
+    "\n",
+    "- ``--path`` is used to create rule file at path given in ``--path``, Default is ``fixit/rules/new.py``\n",
+    "- ``--name`` is used to assign the name of the rule and should be in snake case, Default is name of the rule file if ``path`` provided else `new`. Otherwise, considers the value specified in ``--name``\n",
```
[the remark this answers is not in the document]
Comment (acharles7) [reply docs/source/build_a_lint_rule.ipynb]: done

**Pull request merged**